### PR TITLE
sql: add an event log to sql execution

### PR DESCRIPTION
**Pull request by petermattis**
Visit /debug/events to see the event logs.

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/5374)

### CODE REVIEW

**Comment by tamird** [comment]
:lgtm:

---

Reviewed 5 of 5 files at r1, 4 of 4 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5374#-:-KD9CtkexbI2w4-MFz9K:1417088721)*

File: sql/pgwire/v3.go [hunk 1]
```diff
@@ -111,6 +114,7 @@ type v3Conn struct {
 	doingExtendedQueryMessage, ignoreTillSync bool
 
 	metrics *serverMetrics
+	tr      trace.Trace
```
Comment (tschottdorf) [inline sql/pgwire/v3.go]: unused?

**Comment by tschottdorf** [comment]
Any benchmark impact? This is always-on, right?

File: sql/pgwire/v3.go (hunk 1 above)
Comment (petermattis) [inline sql/pgwire/v3.go]: Detritus. Removed.

**Comment by petermattis** [comment]
@tschottdorf Good point about checking the benchmarks. Yes, this is always on. Turns out `trace.EventLog` is pretty allocation happy. I switched to using `trace.Trace` which is usually used for requests but is used by grpc for long-running streams. There is a small effect on the scan benchmarks:

```
name                   old time/op    new time/op    delta
Scan1_Cockroach-32        192µs ± 1%     194µs ± 2%  +1.23%  (p=0.007 n=10+10)
Scan10_Cockroach-32       240µs ± 3%     246µs ± 3%  +2.49%  (p=0.002 n=10+10)
Scan100_Cockroach-32      621µs ± 2%     631µs ± 3%    ~     (p=0.052 n=10+10)
Scan1000_Cockroach-32    4.18ms ± 2%    4.12ms ± 1%  -1.28%   (p=0.004 n=10+8)

name                   old allocs/op  new allocs/op  delta
Scan1_Cockroach-32          216 ± 0%       222 ± 0%  +2.74%  (p=0.000 n=10+10)
Scan10_Cockroach-32         296 ± 0%       302 ± 0%  +2.03%    (p=0.000 n=8+8)
Scan100_Cockroach-32      1.02k ± 0%     1.03k ± 0%  +0.59%    (p=0.000 n=9+7)
Scan1000_Cockroach-32     8.24k ± 0%     8.24k ± 0%  +0.06%  (p=0.000 n=10+10)
```

I think I can get this down a little more.

**Comment by tschottdorf** [comment]
What about not always-on? Won't the event log of this be extremely spammy
anyway (I have no clear picture of what the output would look like).

**Comment by petermattis** [comment]
screen shot 2016-03-18 at 1 31 55 pm

**Comment by tschottdorf** [comment]
Ah, that does look neat, though it does seem that it should just use a real
trace (which also feeds into net/trace). We should be able to do that when
@paperstreet's context plumbing is done. Until then this is good.

**Comment by petermattis** [comment]
@tschottdorf Won't an (open-tracing) trace be associated with individual requests? These traces are associated with a pgwire connection. With a few changes, I got the tracing down to a single additional allocation in the success case. The overhead is now minimal.

**Comment by tschottdorf** [comment]
No, it builds a tree out of spans - pgwire will have a root span, individual requests fork off of it.

**Comment by bdarnell** [comment]
I know it's useful, but this really shouldn't be on by default. It's pretty bad if any user with access to the admin UI can see every query run by any user.

**Comment by petermattis** [comment]
The `x/net/trace` stuff does have a facility for deeming a log message sensitive and redacting it. And there is a hook to decide who gets to view this stuff. By default, only localhost can look at `/debug/{events,requests}`. See https://godoc.org/golang.org/x/net/trace#pkg-variables.

**Comment by petermattis** [comment]
Heh, and we're overriding `trace.AuthRequest`: https://github.com/cockroachdb/cockroach/blob/master/server/admin.go#L61-L64. Seems straightforward to restrict sensitive data to localhost as the default `AuthRequest` function does.

**Pull request merged**